R/finalize_explanation.R

```diff
@@ -153,3 +168,238 @@ compute_shapley_new <- function(internal, dt_vS) {
+  # function v(S) = E_{\hat{p}(X_sbar | X_s = x_s)} [f(X_sbar, x_s) | X_s = x_s)].
+  dt_squared_diff_original <- sweep(vS, 2, p)^2
+
+  # Check if we are doing weighted mean with the Shapley kernel weights as the weights or using uniform weights
+  if (use_shapley_weights) {
+    if (!exclude_empty_grand_coalitions) {
+      warning(paste0("We do not reccomend seting `exclude_empty_grand_coalitions = FALSE` when ",
+                     "`use_shapley_weights = TRUE`, as the weights of the empty and full coalitions ",
+                     "drastically outweights the other coalitions. This causes the MSEv criterion to be ",
+                     "more or less identical for all approaches."))
+    }
+
+    # We are using the weighted
+    averaging_weights = internal$objects$X$shapley_weight
+
+    # Extract the relevant weights
+    averaging_weights = averaging_weights[id_combination_numbers]
+
+    # Normalize the weights such that the sum to 1
+    averaging_weights = averaging_weights / sum(averaging_weights)
+
+    # Update `dt_squared_diff` matrix by scaling each column by the `averaging_weights`.
+    # Here, one column represents the v(s) estimates for one explicand.
+    dt_squared_diff = dt_squared_diff_original * averaging_weights
+
+    # Compute the mean squared error for each observation, i.e., only averaged over the coalitions.
+    # We take the sum as the weights sum to 1, so denominator is 1.
+    MSEv_eval_crit_explicand <- colSums(dt_squared_diff)
+
+    # Compute the mean squared error for each coalition, i.e., only averaged over the explicands.
+    # Note that here each row is now scaled with its corresponding Shapley kernel weight.
+    MSEv_eval_crit_combination <- rowMeans(dt_squared_diff)
+    MSEv_eval_crit_combination_sd <- apply(dt_squared_diff, 1, sd)
+
+    # Compute the overall weighted squared error averaged over both the coalitions and explicands.
+    MSEv_eval_crit <- mean(dt_squared_diff)
+
+  } else {
+    # We are not using weighted mean
+    dt_squared_diff = dt_squared_diff_original
+
+    # Compute the mean squared error for each observation, i.e., only averaged over the coalitions.
+    MSEv_eval_crit_explicand <- colMeans(dt_squared_diff)
+
+    # Compute the mean squared error for each coalition, i.e., only averaged over the explicands.
+    MSEv_eval_crit_combination <- rowMeans(dt_squared_diff)
+    MSEv_eval_crit_combination_sd <- apply(dt_squared_diff, 1, sd)
+
+    # Compute the overall mean squared error averaged over both the coalitions and explicands.
+    MSEv_eval_crit <- mean(dt_squared_diff)
+  }
+
+  # Set the names entires in the array
+  names(MSEv_eval_crit_explicand) <- paste0("id_", seq(n_explain))
+  names(MSEv_eval_crit_combination) <- paste0("id_combination_", id_combination_numbers)
+  names(MSEv_eval_crit_combination_sd) <- paste0("id_combination_", id_combination_numbers)
+
+  # If we are to return the results as data.table, then we overwrite the previous results.
```

Comment:
Not use what you mean by this, or if these is an issue by always returning the results as a data.table (or always a list). Do we need 2 options here?

Reply:
They were a list first, but then I wanted to add some extra information. E.g `features` for `MSEv_eval_crit_combination` in case we are using sampling of coalitions since it is then nice to see which coalitions `id_combination` number `x` belongs to. I have removed the option and always return a list of `data.tables`.